File: backend/src/main/java/org/project/backend/service/HotelService.java
package org.project.backend.service;

import lombok.RequiredArgsConstructor;
import org.project.backend.dto.hotels.HotelRequest;
import org.project.backend.dto.hotels.HotelResponse;
import org.project.backend.exception.DuplicateResourceException;
import org.project.backend.exception.ResourceNotFoundException;
import org.project.backend.mapper.HotelMapper;
import org.project.backend.model.Hotel;
import org.project.backend.repository.HotelRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class HotelService {

    private final HotelRepository hotelRepository;
    private final HotelMapper hotelMapper;

    @Transactional
    public HotelResponse create(HotelRequest hotelRequest) {
        if (hotelRepository.existsByNomAndCityIdCity(hotelRequest.getNom(), hotelRequest.getCityId())) {
            throw new DuplicateResourceException("Hotel", "nom", hotelRequest.getNom() + " dans cette ville");
        }

        Hotel hotel = hotelMapper.toEntity(hotelRequest);
        Hotel savedHotel = hotelRepository.save(hotel);
        return hotelMapper.toResponse(savedHotel);
    }

    @Transactional
    public HotelResponse update(Integer id, HotelRequest hotelRequest) {
        Hotel hotel = hotelRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Hotel", "id", id));

        if (!hotel.getNom().equals(hotelRequest.getNom()) &&
                hotelRepository.existsByNomAndCityIdCity(hotelRequest.getNom(), hotelRequest.getCityId())) {
            throw new DuplicateResourceException("Hotel", "nom", hotelRequest.getNom() + " dans cette ville");
        }

        hotelMapper.updateEntityFromRequest(hotel, hotelRequest);
        Hotel updatedHotel = hotelRepository.save(hotel);
        return hotelMapper.toResponse(updatedHotel);
    }

    @Transactional(readOnly = true)
    public List<HotelResponse> getAll() {
        return hotelRepository.findAll()
                .stream()
                .map(hotelMapper::toResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public HotelResponse getById(Integer id) {
        Hotel hotel = hotelRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Hotel", "id", id));
        return hotelMapper.toResponse(hotel);
    }

    @Transactional(readOnly = true)
    public List<HotelResponse> getByCity(Integer cityId) {
        return hotelRepository.findByCityIdCity(cityId)
                .stream()
                .map(hotelMapper::toResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public void delete(Integer id) {
        if (!hotelRepository.existsById(id)) {
            throw new ResourceNotFoundException("Hotel", "id", id);
        }
        hotelRepository.deleteById(id);
    }

    @Transactional(readOnly = true)
    public long count() {
        return hotelRepository.count();
    }
}

File: frontend/src/services/hotelService.js
import axiosInstance from "../core/api/axiosInstance";

const hotelService = {
  getAll: async () => {
    const res = await axiosInstance.get("/hotels");
    return res.data;
  },

  getById: async (id) => {
    const res = await axiosInstance.get(`/hotels/${id}`);
    return res.data;
  },

  getByCity: async (cityId) => {
    const res = await axiosInstance.get(`/hotels/city/${cityId}`);
    return res.data;
  },

  create: async (payload) => {
    const res = await axiosInstance.post("/hotels", payload);
    return res.data;
  },

  update: async (id, payload) => {
    const res = await axiosInstance.put(`/hotels/${id}`, payload);
    return res.data;
  },

  delete: async (id) => {
    const res = await axiosInstance.delete(`/hotels/${id}`);
    return res.data;
  },

  count: async () => {
    const res = await axiosInstance.get("/hotels/count");
    return res.data.count;
  },
};

export default hotelService;
